### ISSUE TITLE
0.13.2.AP Pre-Release does not display Linux partitions

### ISSUE THREAD

**Issue by MarioG-X** [dakanji/RefindPlus#106]
### RefindPlus Version

Build of Current Unreleased Code

### Device Type

Apple Mac

### Problem Description

X222 does not display ext4 and btrfs partitions. It displays macOS volumes and boots at least one macOS volume.

### Problem Point

Before starting a loader or tool

### Affected Items

LinOS Loader

### Debug Log

[22h02s3529.log](https://github.com/dakanji/RefindPlus/files/8402969/22h02s3529.log)

### Additional Context

Mac mini 2014. The partitions and volumes are:

Suse153bum Opensuse BTRFS on HD
Suse153bu Opensuse BTRFS on HD
ElementaryOS ext4 on HD

Monterey Mini Container 1 on HD
Monterey Ubiquiti Container 2 on HD

Catalina on SSD   All in same container
Big Sur on SSD
Monterey on SSD

**Comment by dakanji** [comment]
Thanks for the report.

Will come back to this in a while. Can you update to the discussion regarding REL and DBG?

**Comment by dakanji** [comment]
Okay. Got some feedback from @startergo on the REL file thing. Would be great if you would corroborate when you have a moment.

As for this issue, try pressing the `ESC` button to trigger a rescan and see if they show up.

**Comment by MarioG-X** [comment]
I am not sure what you are asking me to do, so I did this:
1. Booted from DBG and tried ESC a few times, nothing changed, log attached.
2. Booted from REL and menu displayed same as DBG, ESC same as DBG, nothing changed.
[22h03q3800.log](https://github.com/dakanji/RefindPlus/files/8404754/22h03q3800.log)

**Comment by dakanji** [comment]
Thanks.

`ESC` only needs to be pressed once for the effect if it would have worked.

On the REL file thing, what I was asking about seems to have been lost in transmission, but all good as you provided the answer in the end.

Back to this, please try X223.
[X223-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8404794/X223-BOOTx64.zip)


Cheers

**Comment by MarioG-X** [comment]
I didn't test everything I had, but I tested a lot, looking good!

Mac mini 2014, almost all partitions, all booted fine, attached one log if needed:
Linux on BTRFS HD
Linux on EXT4 HD
macOS Monterey on HD
macOS Monterey on SSD
[22h03s2027.log](https://github.com/dakanji/RefindPlus/files/8405017/22h03s2027.log)

Imac 2012, almost all partitions, all booted fine:
Linux on BTRFS HD
Linux on BTRFS SSD
macOS Catalina on HD
macOS Catalina on SSD
External disk booted Catalina fine

**Comment by dakanji** [comment]
Great, thanks!

What had happened was that some loading speed optimisation went in that resulted in the issue for those such as you, running OS with non-native FS on Macs, although it didn't seem to affect UEFI Windows. Well, it also did but was not noticeable from a cursory look (only a failure to get the disk name) and was still bootable.

Will continue the tidy up and push the fix at some point

**Comment by dakanji** [comment]
As an aside, how come you have EXT4 and BTRFS installations on partitions with the Apple HFS+ Partition Type GUID?
Is that from how you formatted the disks before installation?

**Comment by MarioG-X** [comment]
It could be that when I layed out the partitions I set aside space for the Linux ones using macOS Disk Uility HPFS. Then once Linux was running I formatted them using Linux. macOS will not allow a blank area when first using Disk Utility. Doing it this way also prevents issues later with macOS. When I tried using Gparted or Linux to delete and recreate partitions I had issues. That was years ago and I don't know if anything changed.  BTW, it also allows them to show up in Disk Utility as "disk1s9, etc" which makes it easy to reclaim into APFS later if needed.

**Comment by dakanji** [comment]
> the PARTITION TYPE appears as BTRFS and EXT4 correctly.

Yes, it does in the last (working) log.

**Comment by MarioG-X** [comment]
BTW, I noticed the display is no longer in sequence by OS type, but now in the order of disks, which _I like much better_.
Before:                                     Linux HD, Linux HD, Linux SSD, macOS HD, macOS SSD.
Now (better for me anyway): Linux HD, Linux HD, macOS HD, Linux SSD, macOS SSD.

**Comment by dakanji** [comment]
Most likely any changes were in the firmware as the scan is based on a firmware defined order of handles returned after a standard `SystemTable->BootServices->LocateHandleBuffer` call.

**Comment by MarioG-X** [comment]
Definitely no firmware changes on 2 old computers in the last day. I just tested the previous RP which displays the old way, so the two RPs display different sequences for sure. Better now for me.

**Comment by dakanji** [comment]
> Definitely no firmware changes on 2 old computers

It might not be the reason for the difference since you say the older version works differently, but note that machines that support Catalina, which includes your 2012 iMac, still receive firmware updates.

This includes a firmware update that came with Security Update 2022-003 of 14 Mar 2022.

**Comment by dakanji** [comment]
Took a look back at changes made and I did change some GUID Constants that impact on DiskIO handles here: https://github.com/dakanji/RefindPlus/commit/0e8f213e1d603c2a2b28af3fd5d33b6ef5d4ef47

RP inherited what looked like a redundant a duplication of these from rEFInd (local duplicates of originals) and while cleaning some stuff up, I decided to ditch the local instances for the originals.

Now wondering why they where duplicated and perhaps it was to compile on GNU-EFI (RP currently only uses EDK2).

It was done while doing other stuff and should have been a separate commit really.
Will have to remember that if/when GNU-EFI builds are ever looked at.
